Library/Homebrew/dev-cmd/pull.rb

```diff
@@ -145,22 +145,37 @@ def pull
         opoo "Current branch is #{branch}: do you need to pull inside master?"
       end
 
-      patch_puller = PatchPuller.new(url, args)
-      patch_puller.fetch_patch
-      patch_changes = files_changed_in_patch(patch_puller.patchpath, tap)
-
-      is_bumpable = patch_changes[:formulae].length == 1 && patch_changes[:others].empty?
-      if do_bump
-        odie "No changed formulae found to bump" if patch_changes[:formulae].empty?
-        if patch_changes[:formulae].length > 1
-          odie "Can only bump one changed formula; bumped #{patch_changes[:formulae]}"
+      patch_url = url + ".patch"
+      issue = patch_url[/([0-9]+)\.patch$/, 1]
+      safe_system "git", "fetch", "--quiet", "origin", "pull/#{issue}/head"
+      if Utils.popen_read("git", "rev-list", "--parents", "-n1", "FETCH_HEAD").count(" ") > 1
+        # Used by forks of homebrew-core that use merge-commits
+        ohai "Fast-forwarding to the merge commit"
+        test_bot_origin = patch_url[%r{(https://github\.com/[\w-]+/[\w-]+)/compare/}, 1]
+        safe_system "git", "fetch", "--quiet", test_bot_origin, "pr-#{issue}" if test_bot_origin
+        system "git", "merge", "--quiet", "--ff-only", "--no-edit", "FETCH_HEAD"
+        unless $CHILD_STATUS.success?
+          safe_system "git", "reset", "--hard", "FETCH_HEAD"
+          odie "Not possible to fast-forward"
         end
-        odie "Can not bump if non-formula files are changed" unless patch_changes[:others].empty?
-      end
-      if is_bumpable
-        old_versions = current_versions_from_info_external(patch_changes[:formulae].first)
+      else
+        patch_puller = PatchPuller.new(url, args)
```

Comment:
Would like to have some of the above logic live in `PatchPuller` and generally share more code with that.